File: apps/web/src/lib/types/task.ts
import type { PillarType, ProjectSourceType, TaskPriority, TaskStatus } from "./enums";

export interface Task {
  id: string;
  title: string;
  description: string | null;
  product_id: string;
  assignee_id: string | null;
  status: string | null;
  priority: string | null;
  pillar: string | null;
  due_date: string | null;
  sort_order: number | null;
  estimated_hours: number | null;
  generation_source: string | null;
  claude_code_prompt: string | null;
  domain_group: string | null;
  phase: string | null;
  is_draft: boolean;
  approved_by: string | null;
  approved_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface TaskTemplate {
  id: string;
  title: string;
  description: string | null;
  pillar: string;
  priority: string | null;
  default_status: string | null;
  source_type: ProjectSourceType;
  order_index: number | null;
  is_active: boolean | null;
  group_id: string | null;
  created_at: string | null;
  updated_at: string | null;
}

export interface TaskTemplateGroup {
  id: string;
  name: string;
  description: string | null;
  source_type: ProjectSourceType;
  is_active: boolean | null;
  order_index: number | null;
  item_count: number;
  created_at: string;
  updated_at: string;
}

export interface TaskTemplateGroupDetail extends TaskTemplateGroup {
  items: TaskTemplate[];
}

export interface KanbanTask {
  id: string;
  title: string;
  description?: string;
  pillar: PillarType;
  priority: TaskPriority;
  status: TaskStatus;
  assignee?: string;
  assigneeId?: string;
  productId?: string;
  dueDate?: string;
  createdAt: string;
}

export interface KanbanColumn {
  id: TaskStatus;
  title: string;
  tasks: KanbanTask[];
}

export const PILLAR_ORDER: PillarType[] = ["development", "product", "business", "marketing"];

File: apps/api/models/task.py
"""Task-related models: tasks, task_templates, task_template_groups."""

import uuid
from datetime import datetime
from typing import TYPE_CHECKING, Optional

from sqlalchemy import Boolean, DateTime, Float, ForeignKey, Integer, String, Text, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from packages.common.db.base import Base, TimestampMixin, UUIDMixin

if TYPE_CHECKING:
    pass


class Task(Base, UUIDMixin, TimestampMixin):
    __tablename__ = "tasks"

    product_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("products.id"), nullable=False
    )
    assignee_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("profiles.id"), nullable=True
    )
    created_by: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("profiles.id"), nullable=True
    )
    title: Mapped[str] = mapped_column(String, nullable=False)
    description: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    status: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    priority: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    pillar: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    due_date: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    sort_order: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    estimated_hours: Mapped[Optional[float]] = mapped_column(Float, nullable=True)
    generation_source: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    claude_code_prompt: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    domain_group: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    phase: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    is_draft: Mapped[bool] = mapped_column(
        Boolean, default=False, server_default="false", nullable=False
    )
    approved_by: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("profiles.id"), nullable=True
    )
    approved_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True), nullable=True
    )


class TaskTemplateGroup(Base, UUIDMixin, TimestampMixin):
    __tablename__ = "task_template_groups"

    name: Mapped[str] = mapped_column(String, nullable=False)
    description: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    source_type: Mapped[str] = mapped_column(String, nullable=False)
    is_active: Mapped[Optional[bool]] = mapped_column(Boolean, default=True, nullable=True)
    order_index: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)

    items: Mapped[list["TaskTemplate"]] = relationship(
        "TaskTemplate",
        back_populates="group",
        order_by="TaskTemplate.order_index",
    )


class TaskTemplate(Base, UUIDMixin):
    __tablename__ = "task_templates"

    group_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("task_template_groups.id", ondelete="SET NULL"),
        nullable=True,
    )
    title: Mapped[str] = mapped_column(String, nullable=False)
    description: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    pillar: Mapped[str] = mapped_column(String, nullable=False)
    priority: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    default_status: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    source_type: Mapped[str] = mapped_column(String, nullable=False)
    order_index: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    is_active: Mapped[Optional[bool]] = mapped_column(Boolean, nullable=True)
    created_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=True
    )
    updated_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=True
    )

    group: Mapped[Optional["TaskTemplateGroup"]] = relationship(
        "TaskTemplateGroup", back_populates="items"
    )
